### PR TITLE
DBZ-191 support parsing default value

### PR DESCRIPTION
**Pull request by sweat123**
https://issues.jboss.org/browse/DBZ-191

### CODE REVIEW

**Comment by ian-axelrod** [comment]
Hiya,

Any chance you can support this for PostgreSQL, as well? It suffers from the same issue.

**Comment by jpechane** [comment]
@ian-axelrod Could you please raise a separate Jira for it? This will need to be done in different way as we would need to include chanec into decoder plugins too.

**Comment by gunnarmorling** [comment]
Hey @sweat123 getting back to this one (sorry for the long delay, I was on PTO/conference travel). I'll review this one ASAP. What's the name we should add for you to the authors list, "Lao Mei"?

@jpechane, I think you took a look already, right? What's your take, any bigger issues remaining? On a related note, are default values supported by Roman's new parser implementation already?

**Comment by gunnarmorling** [comment]
Hey @ian-axelrod, did you get a chance to file that JIRA issue for adding this to the PG connector? Also, I'm curious why you need the default value in the schema. Actual messages should always contain the value, so I'm wondering when the schema's default value would be applied. That's not to say I'm not for implementing this, I'm just eager to learn more about the use case. Thanks!

**Comment by burdiyan** [comment]
Last @gunnarmorling question was already answered in the corresponding JIRA ticket, but for those who are not following and are wondering here's the link: https://issues.jboss.org/browse/DBZ-191?focusedCommentId=13581673&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-13581673. Just FYI.

**Comment by ian-axelrod** [comment]
@gunnarmorling I have not filed it. I do not think I will have a chance to contribute myself, as I only have weekends available and I do not desire to code over the weekend when I already code 6-12 hours 5 days a week. I love coding, but I am not obsessed, hehe. I will probably have to look into a paid solution that sees more frequent updates, sadly.

I can file it so that others can take it on, however. Is that still needed?

**Comment by gunnarmorling** [comment]
@ian-axelrod, totally understandable regarding finding the time for contributing. Everything helps, so if you just filed that JIRA issue for your requirement, that'd be great already.

In terms of frequent updates, we aim at a 3-weekly release cadence, which was interrupted recently a bit due to me being on leave and conference travel for a while. The change here (for MySQL) is scheduled for 0.8 which is due in two weeks.

As far as paid support is concerned, can you please send me an e-mail to "myfirstname AT hibernate.org" to discuss that topic? Cheers!

File: debezium-connector-mysql/src/test/resources/ddl/default_value.sql [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+  D DATETIME(1) DEFAULT '2018-01-03 00:00:10.7',
+  E DATETIME(6) DEFAULT '2018-01-03 00:00:10.123456',
+  F YEAR NOT NULL DEFAULT 1,
+  G TIME DEFAULT '00:00:00',
+  H TIME(1) DEFAULT '23:00:00.7',
+  I TIME(6) DEFAULT '23:00:00.123456',
+  J TIMESTAMP DEFAULT CURRENT_TIMESTAMP
+);
+INSERT INTO DATE_TIME_TABLE
+VALUES (DEFAULT, DEFAULT, DEFAULT, DEFAULT, DEFAULT, DEFAULT, DEFAULT, DEFAULT, DEFAULT, DEFAULT );
```
Comment (gunnarmorling) [inline debezium-connector-mysql/src/test/resources/ddl/default_value.sql]: Could you add a linebreak at EOF?

**Comment by gunnarmorling** [comment]
Thanks, @burdiyan, that makes a lot of sense. If you have any thoughts on that timestamp issue (function-based default value), please let us know.

**Comment by burdiyan** [comment]
@gunnarmorling Oh, I haven't thought about the case of timestamps default VALUE. I think to avoid compatibility issues with the schema registry in this particular case the default value on the schema could be 0/empty string (I don't remember if TIMESTAMP is parsed to string or long from the top of my head now).

But basically, the idea is to put something in the schema to avoid issues on the schema registry side, since anyway the correct default value will be set by the database itself, so in theory there's nothing to be worrying about.

**Comment by gunnarmorling** [comment]
Yeah, that could be a way out. Probably would have to be something like epoch then (1970-1-1 00:00:00). Not ideal, but as you say, if documented well, it may be good enough. Let's see what @sweat123 thinks about it and then decide.

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlDdlParser.java [hunk 2]
```diff
@@ -1632,6 +1670,32 @@ protected void parseDefaultClause(Marker start) {
...
+     */
+    protected void parseDefaultClause(Marker start, ColumnEditor column) {
+        tokens.consume("DEFAULT");
+        if (isNextTokenQuotedIdentifier()) {
+            // We know that it is a quoted literal ...
+            Object defaultValue = parseLiteral(start);
+            column.defaultValue(defaultValue);
+        } else {
+            if (tokens.matchesAnyOf("CURRENT_TIMESTAMP", "NOW")) {
+                parseCurrentTimestampOrNow();
```
[the remark this answers is not in the document]
Comment (sweat123) [reply debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlDdlParser.java]: @gunnarmorling em. `CURRENT_TIMESTAMP` and `NOW` will be replaced with epoch timestamp (1970-01-01 00:00:00) now :smile:;

**Comment by jpechane** [comment]
@gunnarmorling It should not be difficult to merge this PR with the new parser. I think the PR looks great but not all datatypes are sill supported e.g. geometry.
We should either support all types or enumerate those supported to avoid confusion between users.

**Comment by gunnarmorling** [comment]
@jpechane `GEOMETRY` columns don't support default values [as per the docs](https://dev.mysql.com/doc/refman/8.0/en/data-type-defaults.html) (as JSON, BLOB and TEXT). Do you see any specific types still unsupported?

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlDdlParser.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (gunnarmorling) [reply debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlDdlParser.java]: Ok, cool. Thanks for the update!

**Comment by jpechane** [comment]
@gunnarmorling Thanks for pointing it out to me! I went through the rest of datattypes and IMHO we are clear to go.

**Comment by gunnarmorling** [comment]
Thanks for double-checking. It looks good to me, too. Going to merge it and
add a sentence to the docs about the default values being considered in
schemas.

**Comment by gunnarmorling** [comment]
Rebased and applied. Thanks a lot, @sweat123! I added the name "Luobo" to _COPYRIGHT.txt_ (taken from the commit authorship). Please let me know if I should use something else instead.

Great contribution overall, much appreciated!

**Comment by sweat123** [comment]
@gunnarmorling Thanks for the review! 
My name is `WenZe Hu`, thanks a lot! :smile:

**Pull request closed without merging**